File: Backtracking_and_Search/LC_78_Subsets.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.19

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans;
        int n = nums.size();
        for (int l = 0; l <= n; l++){
            vector<int> curr;
            dfs(nums, 0, l, curr, ans);
        }
        return ans;
    }
private:
    void dfs(vector<int>& nums, int start, int length, vector<int>& curr, vector<vector<int>> & ans){
        if (curr.size() == length){
            ans.push_back(curr);
            return;
        }
        for(int i = start; i < nums.size(); i++){
            curr.push_back(nums[i]);
            dfs(nums, i+1, length, curr, ans);
            curr.pop_back();
        }
    }
};

File: Backtracking_and_Search/LC_78_Subsets.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.19

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        # even cleaner implementation
        ans = []
        n = len(nums)
        def dfs(i, curr):
            if i == n:
                ans.append(curr[:])
                return
            # include
            curr.append(nums[i])
            dfs(i+1, curr)
            # not include
            curr.pop()
            dfs(i+1, curr)
        
        dfs(0, [])
        return ans
        """
        # generate all combinations for the given length
        ans = []
        n = len(nums)
        def dfs(l, nums, start, curr, ans):
            if len(curr) == l:
                ans.append(curr[:])
                return 
            for i in range(start, len(nums)):
                curr.append(nums[i])
                dfs(l, nums, i+1, curr, ans)
                curr.pop()
                
        for l in range(n+1):
            dfs(l, nums, 0, [], ans)
        return ans
        """
        